components/suggest/src/query.rs

```diff
@@ -143,43 +143,84 @@ impl SuggestionQuery {
-        let keywords = self.parse_keywords();
+pub struct FtsQuery<'a> {
+    pub sqlite_match: String,
+    pub prefix_match: bool,
+    keywords: Vec<&'a str>,
+}
+
+impl<'a> FtsQuery<'a> {
+    fn new(keyword: &'a str) -> Self {
+        // Parse the `keyword` field into a set of keywords.
+        //
+        // This is used when passing the keywords into an FTS search.  It:
+        //   - Strips out any `():^*"` chars.  These are typically used for advanced searches, which
+        //     we don't support and it would be weird to only support for FTS searches, which
+        //     currently means Fakespot searches.
+        //   - splits on whitespace to get a list of individual keywords
+        let keywords = Self::split_terms(keyword);
         if keywords.is_empty() {
-            return String::from(r#""""#);
+            return Self {
+                keywords,
+                sqlite_match: String::from(r#""""#),
+                prefix_match: false,
+            };
         }
         // Quote each term from `query` and join them together
-        let mut fts_query = keywords
+        let mut sqlite_match = keywords
             .iter()
             .map(|keyword| format!(r#""{keyword}""#))
             .collect::<Vec<_>>()
             .join(" ");
         // If the input is > 3 characters, and there's no whitespace at the end.
         // We want to append a `*` char to the end to do a prefix match on it.
         let total_chars = keywords.iter().fold(0, |count, s| count + s.len());
-        let query_ends_in_whitespace = self.keyword.ends_with(' ');
-        if (total_chars > 3) && !query_ends_in_whitespace {
-            fts_query.push('*');
+        let query_ends_in_whitespace = keyword.ends_with(' ');
+        let prefix_match = (total_chars > 3) && !query_ends_in_whitespace;
+        if prefix_match {
+            sqlite_match.push('*');
+        }
+        Self {
+            keywords,
+            prefix_match,
+            sqlite_match,
         }
-        fts_query
+    }
+
+    pub fn sqlite_match_without_prefix_match(&self) -> &str {
+        self.sqlite_match
+            .strip_suffix('*')
+            .unwrap_or(&self.sqlite_match)
+    }
+
+    pub fn match_required_stemming(&self, title: &str) -> bool {
+        let title = title.to_lowercase();
+        let split_title = Self::split_terms(&title);
+
+        !self.keywords.iter().enumerate().all(|(i, keyword)| {
+            split_title.iter().any(|title_word| {
+                let last_keyword = i == self.keywords.len() - 1;
+                let value = if last_keyword && self.prefix_match {
+                    title_word.starts_with(keyword)
+                } else {
+                    title_word == keyword
+                };
+                value
+            })
+        })
+    }
+
+    fn split_terms(phrase: &str) -> Vec<&str> {
+        phrase
+            .split([' ', '(', ')', ':', '^', '*', '"', ','])
```

Comment:
I added the `,` char in the list of things to split on.  This way the comma in `Trail Running,` isn't included in the search terms.  The FTS tokenizer ignores that, but it was messing up the stemming check logic.